docs/design/prerequisites-and-targets.md

```diff
@@ -0,0 +1,117 @@
+because refreshing consults the state rather than running the program, so it will proceed with stale
+credentials and likely fail authentication.
+
+Usually it is possible to work around this by using `pulumi up -t <provider URN>` to rerun the
+program and update the inputs to the provider, before running the refresh operation. However, this
+workaround isn't available when using the operator, because there is no representation of "run this
+before" or targeted updates in its API.
+
+## Design
+
+This design involves two additions to the API:
+
+ - "prerequisites", a list of other stack objects that must be run before the stack under
+   consideration is run; and,
+ - "targets", a list of URNs to be given to the Pulumi engine when running the stack.
+
+### Prerequisites
+
+The field `.spec.prerequisites` gives a list of references to stacks. The controller makes sure each
+has been run recently before running the stack under consideration. This means, for instance, you
+can update credentials in the state before re-running a program that uses those credentials.
+
+Usually it will be important that a prerequisite has run recently. The field
+`.spec.prerequisite[].updatedWithinDuration` puts a limit on how long ago it was last run:
+
+```
+spec:
+  prerequisites:
+  - stackRef:
+      name: update-credentials
+    unless:
+      updatedWithinDuration: 1m
+```
+
+If the referenced stack has not been run in the last `updatedWithinDuration` duration, it will be
+queued, and the current stack requeued.
+
+**Q. How does the controller queue the prerequisites?**
+
+One mechanism is to simply add an annotation to the object, which will trigger an update event and
+notify the controller (though the controller would have to be adapted to not filter those
+changes).
+
+This mechanism is used neatly by Flux to queue dependent objects; the CLI also employs it to make
+interactions synchronous (mark an object with the annotation, then wait for its status to report
+that it was processed).
+
+**Q. Can you refer to the same stack in two Stack objects?**
+
+For this to be useful for updating stale state, the depended-upon Stack must update the state used
+by the depending Stack (or the program must be rewritten to use stack references, which arguably
+would be good practice anyway). This would require it to have the same backend and name the same
+stack; this is likely to be possible, but may cause conflicts if not carefully managed.
+
+Note that it is not _always_ a requirement that a prerequisite uses the same stack or backend --
+that is how you use it to e.g., update rotated credentials, if that's done all in one program.
+
+### Targets
+
+The field `.spec.targets` has a list of URNs to supply to the Pulumi automation API when running the
+stack.
+
+## How this design solves the stale provider state problem
+
+To solve the stale provider state problem, you would make two stacks:
```

Comment:
One challenge - whether here or through one of the alternatives discussed later is to know ahead of time the URNs to target. Making this a prerequisite for the regular stack execution is going to introduce a circular dependency no? As in - the provider URN to target doesn't exist until the full stack is executed?

Reply:
Yes -- this is a problem for this workaround, wherever you do it. I can think of a few improvements:

 - instruct people to make the "main" stack first, extract the URN of the provider, then set up the refresh stack. Since the problem doesn't exist unless you have run the stack at least once, there's no circularity, just a step sideways.
 - ask for something other than URNs, and try to figure out what the URNs will be. This _might_ end up being circular, or at least fiddly.

Reply:
I wonder if it would make sense to model the dependency aspect as a separate CRD - e.g. `stacklink` - something that links dependent stacks together. This essentially looks like a dependency link between stacks and feels a bit awkward to shove this into the stack spec? In my view the prerequisite - and target seem like constraints placed on a stack which can be more dynamic in nature. We may be able to provide a simple expression to resolve URNs etc. over time to make this easier. 

In some ways there are echoes of https://github.com/pulumi/pulumi/issues/2309 here which is a much requested feature and could enable a number of the kind of workloads mentioned there over time.

cc @lukehoban for thoughts as well.

Reply:
> This essentially looks like a dependency link between stacks and feels a bit awkward to shove this into the stack spec?

I don't think it's awkward -- there are already a bunch of fields that are to do with when and how a Stack object is processed (rather than what it executes), like `RetryOnUpdateConflict` and `ContinueResyncOnCommitMatch`. This is just another one of those.

> In some ways there are echoes of https://github.com/pulumi/pulumi/issues/2309

For sure; if `StackReference` dependencies could be chased down and updated, it would give you a natural way to write programs that didn't need the workaround here. I'm not sure it would cover every use case*, but to me it's the right primitive to start with.

*Why not? Hand waving a bit,
 - using a stack reference means you have to hand values from one program to another via outputs, and you may not want intermediate values to be outputs;
 - with a stack reference, you don't get to say how the referenced stack is executed (e.g., you can't provide targets)

Reply:
The main benefit of having another type to represent dependence is that it makes it possible to separate the ownership of the stack definition and its dependence on other stacks. This could be important when crossing responsibility boundaries, e.g., a platform team can make app stacks depend on platform infrastructure, without the app teams needing to know about it.